Handle site connection state in Jetpack plugin connection flow

staskus opened the pull request:
Fixes: #16489
Related to: pe5sF9-B9-p2

## Description

There're 2 types of connections for the Jetpack plugin and the site:
- **Site Connection**. This is what is established by clicking the “Set up Jetpack” button, and before completing the WordPress.com login/auth.
- **User Connection**. This is what is established by completing the WordPress.com login/auth.

**Site Connection** allows to use some of Jetpack's features on the web but is not intended to access any features on mobile, therefore in the app we need to establish **User Connection**.

The issue is that the current Jetpack plugin connection flow does not handle **Site Connection** state:
- On iOS, we continue asking users to "Sign in" to WP.com account even if user has already signed in when accessing such features such as Stats
- On Android, we ask users to "Sign in" to WP.com account but after that Jetpack features such as Stats don't work

### Workaround
There's no workaround for this issue on the app, the issue needs to be fixed on the web by either:
- Establishing **User Connection**
- Disconnecting Jetpack plugin and then doing the fresh installation flow on mobile

## Solution

After investigating the problem, the only way I found to solve the issue was to use Jetpack REST API to fetch the connection URL which then can be loaded to finish the connection.

The example is taken from [WooCommerce](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Authentication/Jetpack%20Setup/Native%20Jetpack%20Setup/JetpackSetupViewModel.swift) which implemented native Jetpack plugin installation and connection flow. However, there're many different states and corner cases to handle to create a native solution that wouldn't be viable in the scope of this bug fix.

I decided to apply the partial native Jetpack installation solution only in the use-case that is not currently handled. This way the currently created flow is not affected in any way and we're only aiming to help the users who are stuck in **site connection** state.

| State      | Current solution | New solution |
| ----------- | ----------- | ----------- |
| Jetpack not installed / not active    | ✅ Uses `jetpack-install/%@/` endpoint and handles the flow in `JetpackRemoteInstallViewController`       | _No change_ |
| Jetpack not connected      | ✅ Opens `https://wordpress.com/jetpack/connect` in WebView and handles redirections to sign in or complete the flow in `JetpackConnectionWebViewController`       | _No change_ |
| Jetpack has site connection   | ❌ Treats Jetpack as fully connected and shows "Sign In" CTA. Logging into WP.com account does not change the state and Jetpack features do not work.        | ✅ Use [Jetpack REST API](https://github.com/Automattic/jetpack/blob/trunk/docs/rest-api.md#get-wp-jsonjetpackv4connectionurl) to fetch the Jetpack Connection URL that is loaded into WebView to complete the connection and sign in process in `JetpackConnectionWebViewController` |
| Jetpack has user connection | ✅ Treats Jetpack as fully connected and shows "Sign In" CTA. Logging into WP.com account enables Jetpack features on the app | _No change_  

## Testing instructions

- Keep in mind known issues: https://github.com/wordpress-mobile/WordPress-iOS/issues/19261
- Try testing on different WP.com accounts (with/without 2FA)

### Case 1: Site connection created on the web
1. On the web, create a self-hosted site, install the Jetpack plugin, activate the Jetpack plugin
2. On the web, go to "Jetpack" section, click "Set up" but do not complete the process by logging into .com account. Now **Site connection** is reached
3. Log into the self-hosted site on the Jetpack app
4. Click Stats/Notifications
5. "Connect your user account" CTA should be visible
6. Perform the connection flow
7. Confirm that Jetpack is installed and now Stats/Notifications could be used

### Case 2: Site connection created on the mobile
1. On the web, create a fresh self-hosted site
3. Log into the self-hosted site on the Jetpack app
5. Click Stats/Notifications
6. "Install Jetpack" CTA should be visible
7. Perform the installation flow
8. When the app prompts to log into WP.com account cancel the flow and come back to Stats view
9. "Connect your user account" CTA should be visible
9. Perform the connection flow
10. Confirm that Jetpack is installed and now Stats/Notifications could be used

### Case 3: Site connection when we log into WP.com account before adding self-hosted site
1. Complete 1-2 steps from Case 1
2. Log into the app using WP.com account
3. Add self-hosted site
4. Click Stats/Notifications
5. "Connect your user account" CTA should be visible
6. Perform the connection flow, it may have a different order since we already have logged into WP.com account in step 2
7. Confirm that Jetpack is installed and now Stats/Notifications could be used for a self-hosted site

### Case 4: Full installation (regression)
1. Create a fresh self-hosted site without the Jetpack plugin installed
2. Log into the self-hosted site on the Jetpack app
5. Click Stats/Notifications
6. "Install Jetpack" CTA should be visible
7. Perform the installation flow and log into WP.com account
10. Confirm that Jetpack is installed and now Stats/Notifications could be used


### Case 5: User Connection without WP.com account on the app
1. Complete Case 4
2. Log out from the app
3. Log in to self-hosted site again
4. Open Stats/Notifications
5. "Sign in" CTA should be visible. Since there's a User Connection established on the web.
6. Complete sign in process
7. Confirm that Jetpack is installed and now Stats/Notifications could be used

## Regression Notes

1. Potential unintended areas of impact

Current Jetpack installation flow

10. What I did to test those areas of impact (or what existing automated tests I relied on)

Lots of manual testing in different scenarios

12. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Images & Videos

### Handle site connection state (Case 1)
https://user-images.githubusercontent.com/4062343/224548495-8991ea14-66a2-4e73-a641-9de0d05a7d51.mp4

### Full installation flow (Case 4)

https://user-images.githubusercontent.com/4062343/224548453-f4fccec7-7a51-491d-9493-0938ea3efe9a.mp4

### Handle user connection state (Case 5)

https://user-images.githubusercontent.com/4062343/224548653-822ce0c1-eff4-4929-8099-7859584e02e3.mp4

staskus committed:
Create JetpackNativeConnectionService that uses Jetpack REST API

Implementing fetchJetpackConnectionURL and fetchJetpackUser methods to support Jetpack connection during site-connection state

staskus committed:
Use alternative Jetpack connection method for site-connection state

Current implementation does not support the site-connection state, when site is connected to Jetpack without connecting to .com account.

In this case use Jetpack REST API to fetch connection URL, load it into webView, and handle completion when the plans page is opened.

staskus committed:
Show connection CTA in JetpackLoginViewController

Handle site-connection state by showing Connection CTA instead of Login CTA in JetpackLoginViewController

staskus committed:
Merge branch 'trunk' into fix/16489-jetpack-connection-flow-does-not-handle-site-connection-state

staskus committed:
Add tracks

staskus committed:
Small code organization improvements

staskus committed:
Allow multiline button title for connection

sla8c commented:
@staskus in most of the cases you have this test step:
 * Stats / Notifications.

Can I confirm if this is the flow you wanted the reviewer to follow?
1.  Tap Stats  
2.  Tap Notifications on the Stats screen

Related screenshots below

![image](https://user-images.githubusercontent.com/88816724/226421853-7e4a1ec8-d7be-4132-9c38-2c9e3d0ec447.png)

![image](https://user-images.githubusercontent.com/88816724/226421910-3dadbdbc-174f-4f07-9483-3a919d17f6ff.png)

staskus commented:
> in most of the cases you have this test step: Stats / Notifications.
> Can I confirm if this is the flow you wanted the reviewer to follow?

Thanks for clarifying, @sla8c!

I meant either test by going to Stats view or Notifications tab since both should show identical Jetpack installation flows.

sla8c commented:
@staskus I ran into a crash on Case 3.  I will DM you the details with stack trace etc.

staskus committed:
Merge branch 'trunk' into fix/16489-jetpack-connection-flow-does-not-handle-site-connection-state

staskus commented:
I can reproduce the crash even with the old unchanged flow. I'm investigating if it's related to https://github.com/wordpress-mobile/WordPress-iOS/pull/20146 which had recent changes in the PlanStorage

staskus committed:
Use context.existingObject to fetch a blog when updating domain credit

Blog.lookup may not return an existing blog since it doesn't exist in the current context.

Use context.existingObject instead which loads an object from the persistent store if it doesn't belong in current context

staskus commented on WordPress/Classes/Services/PlanService.swift:
```diff
@@ -241,15 +241,15 @@ extension PlanService {
-        guard let blog = try? Blog.lookup(withID: siteID, in: context) else {
-            let error = "Tried to update a plan for a non-existing site (ID: \(siteID))"
+    static func updateHasDomainCredit(_ planID: Int, forBlog blog: Blog, hasDomainCredit: Bool, in context: NSManagedObjectContext) {
+        guard let blogInContext = try? context.existingObject(with: blog.objectID) as? Blog else {
```
[Similar solution used in recent refactoring of Sharing Service](https://github.com/wordpress-mobile/WordPress-iOS/blob/157aee0f9d2e9429e50a863ca6d7ecefbafe5be9/WordPress/Classes/Services/SharingService.swift#L144)

staskus committed:
Update PlanService.swift

sla8c commented on WordPress/Classes/ViewRelated/Jetpack/Install/JetpackNativeConnectionService.swift:
```diff
@@ -0,0 +1,87 @@
+            case .success(let json):
+                do {
+                    let data = try JSONSerialization.data(withJSONObject: json)
+                    let jetpacUserData = try JSONDecoder().decode(JetpackUserData.self, from: data)
```
@staskus nit - spelling?  `jetpackUserData`?

staskus replied:
Thanks, will be fixed!

sla8c reviewed (approved):
@staskus LGTM; I made a small comment regarding maybe you want to update the spelling on a variable name.  If you decide to change that I don't think it needs a re-review.

sla8c commented:
@staskus just noticed before you merge it would be best to investigate the failing test(s) thx

staskus commented on WordPress/Classes/ViewRelated/Jetpack/Install/JetpackNativeConnectionService.swift:
```diff
@@ -0,0 +1,87 @@
+            case .success(let json):
+                do {
+                    let data = try JSONSerialization.data(withJSONObject: json)
+                    let jetpacUserData = try JSONDecoder().decode(JetpackUserData.self, from: data)
+                    completion(.success(jetpacUserData.currentUser))
+                } catch {
```
```suggestion
                    let jetpackUserData = try JSONDecoder().decode(JetpackUserData.self, from: data)
                    completion(.success(jetpackUserData.currentUser))
                } catch {
```

staskus committed:
Update misspelling
